PowerPointLabs/PowerPointLabs/PositionsLab/PositionsPaneWPF.xaml.cs

```diff
@@ -1353,114 +1360,122 @@ public void ExecutePositionsAction(Action<List<PPShape>, bool> positionsAction,
-                    SaveSelectedShapePositions(selectedShapes, allShapePos);
+                    // invalid selection!
+                    return;
                 }
-                else
+
+                ShapeRange simulatedShapes = null;
+
+                try
                 {
-                    UndoPreview();
-                    _previewCallBack = null;
-                    this.StartNewUndoEntry();
-                }
+                    ShapeRange selectedShapes = selection.ShapeRange;
 
-                simulatedShapes = DuplicateShapes(selectedShapes);
-                List<PPShape> simulatedPPShapes = ConvertShapeRangeToPPShapeList(simulatedShapes, 1);
-                float[,] initialPositions = SaveOriginalPositions(simulatedPPShapes);
+                    if (isPreview)
+                    {
+                        SaveSelectedShapePositions(selectedShapes, allShapePos);
+                    }
+                    else
+                    {
+                        UndoPreview();
+                        _previewCallBack = null;
+                        this.StartNewUndoEntry();
+                    }
 
-                positionsAction.Invoke(simulatedPPShapes, dimension);
+                    simulatedShapes = DuplicateShapesForDistributeOperation(selectedShapes);
+                    List<PPShape> simulatedPPShapes = ConvertShapeRangeToPPShapeList(simulatedShapes, 1);
+                    float[,] initialPositions = SaveOriginalPositions(simulatedPPShapes);
 
-                SyncShapes(selectedShapes, simulatedPPShapes, initialPositions);
+                    positionsAction.Invoke(simulatedPPShapes, dimension);
 
-                if (isPreview)
-                {
-                    _previewIsExecuted = true;
+                    SyncShapes(selectedShapes, simulatedPPShapes, initialPositions);
+
+                    if (isPreview)
+                    {
+                        _previewIsExecuted = true;
+                    }
                 }
-            }
-            catch (Exception ex)
-            {
-                if (!isPreview)
+                catch (Exception ex)
                 {
-                    ShowErrorMessageBox(ex.Message, ex);
+                    if (!isPreview)
+                    {
+                        ShowErrorMessageBox(ex.Message, ex);
+                    }
                 }
-            }
-            finally
-            {
-                if (simulatedShapes != null)
+                finally
                 {
-                    simulatedShapes.Delete();
-                    GC.Collect();
+                    if (simulatedShapes != null)
+                    {
+                        simulatedShapes.Delete();
+                        GC.Collect();
+                    }
                 }
             }
         }
 
         public void ExecutePositionsAction(Action<List<PPShape>, float, float> positionsAction, float dimension1, float dimension2, bool isPreview)
```

Comment:
Ignoring this as it involves simplifying the method further which is not possible without changing the basic functionality of the current method

Reply:
Do you guys have the permission to edit SonarQube? It may be beneficial to decide which rules should be ignored in our project to avoid such problems, especially for future developers

Reply:
I think we don't need to edit the SonarQube, not that we were given permissions to. This is because the problem in code quality is isolated for only certain functions in PositionsLab, which was code by previous few batches of students. It would be better to maintain the same rules now to maintain the same strictness in code standard in the overall project, and decide which to ignore on a case-by-case basis

Reply:
As long as the current rules in place match with the current code quality. You are right that we can ignore it on a case by case basis for legacy codes.